pupilfirst/pupilfirst
# Issue #1494 Resolved: Potential crash when linking a Discord account that's linked to 100 servers

## Description
This pull request addresses the issue where our system crashes when a user attempts to link their Discord account to our server, and the server count exceeds Discord's limit of 100 servers per user. The crash occurs due to `Discord::AddMemberService` raising `Discordrb::Errors::ServerLimitReached`. This has been raised as an issue by @harigopal (https://github.com/pupilfirst/pupilfirst/issues/1494)

## Proposed Changes
- Modified the error handling within `Discord::AddMemberService` to catch `Discordrb::Errors::ServerLimitReached`.
- Implemented a user-friendly error message to inform the user about the server limit restriction.
- Updated the response mechanism to provide a clear explanation to the user instead of crashing the system and displaying a 500 error.

## How to Test
1. Attempt to link a Discord account that already belongs to 100 servers.
2. Verify that instead of crashing, the system responds with an informative message indicating the reason for failure to link the account.

This change prevents system crashes and provides a better user experience by explaining why their Discord account couldn't be linked or why an invite to the server couldn't be sent.


@pupilfirst/developers

## Merge Checklist

- [x] Add specs that demonstrate bug / test a new feature.
- [ ] Check if route, query, or mutation authorization looks correct.
  - Add tests for authorization, if required.
- [x] Ensure that UI text is kept in I18n files.
- [ ] Update developer and product docs, where applicable.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Check if new tables or columns that have been added need to be handled in the following services:
  - `Users::DeleteAccountService`
  - `Courses::CloneService`
  - `Courses::DeleteService`
  - `Courses::DemoContentService`
  - `Levels::CloneService`
  - `Schools::DeleteService`
- [ ] Check if changes in _packaged_ components have been published to `npm`.
- [ ] Add development seeds for new tables.
- [ ] If the updates involve Graph mutations ensure that the files are migrated to the new approach without a mutator.
- [ ] If the updates involve adding a new table ensure that rate limiting is added and documented in the `docs/developers/rate_limiting.md` file.
